web/docs/internal/keystroke-lifecycle.md

```diff
@@ -0,0 +1,90 @@
+# The Keyman Engine for Web keystroke lifecycle
+
+Depending on the variant of Keyman Engine for Web in play, the engine integrates either with a host app or with the current browser page to interpret and/or intercept keystroke events and produce the desired context manipulations corresponding to them.  While some details diverge between `app/browser` and `app/webview` builds, the general structure and pattern for this process holds.
+
+This document will break down the full keystroke lifecycle into four distinct phases, as follows:
+
+1. Keystroke pre-processing: event interception + interpretation (pre-processing)
+2. Keystroke processing: keystroke main-group rule evaluation
+3. Keystroke post-processing: keystroke post-keystroke evaluation + side-effects
+4. Keystroke resolution: UI state updating
+
+## Keystroke pre-processing
+
+Both the `app/browser` and `app/webview` products are capable of handling simulated keystrokes from the on-screen keyboard as well as hardware keystrokes.  For hardware keystrokes, this varies between the two, while the majority of on-screen keyboard keystrokes are handled identically.
+
+### `app/browser`
+
+The `app/browser` version of the engine is designed to integrate with the active page of the user's browser, giving it far more direct access to the user's true context in addition to direct access to handling DOM events resulting from hardware keystrokes.  Its `HardwareEventKeyboard` class is the primary module for establishing the needed engine event hooks and for interpreting the corresponding DOM events as they arise.
+
+The bulk of DOM key events trigger keystroke processing on key-down, though modifier keys also receive key-up handling.  Key-press events are handled selectively for held backspaces and possibly certain keystrokes for mnemonic keyboards.  In all cases, DOM keyboard event objects are preprocessed by the centralized `preprocessKeyboardEvent` method, which converts the DOM event format into Web's `KeyEvent` format, which is the form required by the keyboard rules within compiled JS keyboards.  This includes mnemonic keystroke preprocessing and remapping in addition to AltGr emulation.
+
+### `app/webview`
+
+The `app/webview` version of the engine, which is designed to be used while embedded in a platform-specific host app, does not have direct access to standard hardware keystroke events.  It's run in a WebView that does not take input - a WebView owned by the host app.  It is the responsibility of the host app to handle hardware keystroke events and preprocess them on behalf of the Web engine, then forward them to the Web engine via a carefully-constructed JS call from the host upon the WebView.  The `PassthroughKeyboard` class within `app/webview` space provides the method `raiseKeyEvent` as an internal API for this purpose, which converts the mobile-app format for hardware keystroke into the internal `KeyEvent` format.  This includes mnemonic keystroke processing and remapping.
```

Comment:
Why "carefully-constructed"? I don't think we need to say that. All code is carefully constructed (I hope!)

It's not totally clear from the sentence "This includes mnemonic keystroke processing and remapping." what "this" refers to. Perhaps say "`raiseKeyEvent` also does the mnemonic..."